In [ ]:
%qtconsole

# Expectation Maximization (theory)
Please see slides 17-30 in `ML-Clustering-L2.pdf` for an explanation of the Expecation Maximization algorithm. Let $D = \{x_1,...,x_n\}\subseteq \mathbb{R}^d$. Recall that in the EM algorithm we represent a clustering as a probability distribution over the points. The probability that a point, $x$, belongs to some cluster $C_i$ is assumed to be Gaussian, i.e.:

$$
\Pr(x\ | \ C_i) = \frac{1}{\sqrt{(2\pi)^d|\Sigma_i|}}\cdot e^{-\frac{1}{2}\cdot (x-\mu_i)^\intercal\cdot (\Sigma_i)^{-1}\cdot (x-\mu_i)}, \ \quad i=1, \dots k \ \qquad (1).
$$

Where $|\Sigma_i |$ denotes the determinant of $\Sigma_i$.



Each cluster $C_i$ is parameterized with a cluster mean $\mu_i\in\Bbb R^d$, a covariance matrix $\Sigma_i\in\Bbb R^{d\times d}$ and a prior probability $\Pr(C_i)$.

<!--Given a clustering $M = \{C_1,\ldots,C_k\}$, we can compute
$$
\Pr(x) = \sum_{i=1}^k \Pr(C_i)\cdot\Pr(x|C_i),
$$
where we estimate $\Pr(C_i)$ by $W_i$, the frequency of the cluster $C_i$ in the data (i.e. the ratio of data points belonging to cluster $C_i$).-->

The Expectation Maximization has two steps: Expectation and Maximization (hence the name). Before these steps we need to initialize a clustering $M = \{C_1,\ldots,C_k\}$. A (very) simple heuristic for initialization is:
- Choose cluster mean $\mu_i\in\Bbb R^d$ uniformly at random (in the adequate region). 
- Initialize the covariance matrix $\Sigma_i\in\Bbb R^{d\times d}$ as the identity matrix. 
- Initialize the prior probability $\Pr(C_i)$ to $\frac{1}{k}$. 


Then we repeat the following two steps until the sum $\sum_{i=1}^k \|\mu_i - \mu'_i\|$ is below some pre-specified threshold $\epsilon$ (where $\mu'_i$ and $\mu_i$ are the means computed in two consecutive executions).

- <b>Update probabilities (expectation step):</b> For all pairs $C_i$ and $x_j$ compute $\Pr(C_i|x_j) = \frac{\Pr(x_j \ | \ C_i)\cdot \Pr(C_i)}{\Pr(x_j)}$. To do this we need to compute $\Pr(x_j \ | \ C_i)$ and $\Pr(x_j)$. Compute $\Pr(x_j \ | \ C_i)$ by the formula (1) above and compute $$\Pr(x_j)= \sum_{i=1}^k \Pr(C_i)\cdot\Pr(x_j \ | \ C_i).$$ 
  
- <b>Update clustering (maximization step): </b>Compute a new model by recomputing $W_i \approx \Pr(C_i)$, $\mu_i$ and $\Sigma_i$ as

$$
W_i = \frac{1}{n}\cdot\sum_{x\in D}\Pr(C_i \ | \ x)\approx P(C_i),
\qquad
\mu_i = \frac{\sum_{x\in D}^n x\Pr(C_i \ | \ x)}{\sum_{x\in D} \Pr(C_i \ | \ x)},
\qquad
\Sigma_i = \frac{\sum_{x\in D} \Pr(C_i \ | \ x)\cdot(x-\mu_i)\cdot(x-\mu_i)^\intercal}{\sum_{x\in D} \Pr(C_i \ | \ x)}.
$$

<b>Question 1:</b> What is the objective function of Expectation Maximization (EM) clustering problem? 

HINT: See slides pages 19. 

<!--<b>Question 1 (b): </b>Rewrite the objective for $\text{dist}(x,y)=||x-y||$-->

<b>Question 2:</b> What are the E-step, M-step? What is estimated, what is fixed?

<b>Question 3:</b> Does EM algorithm provide any guarantees for finding the optimal solution?

HINT: What is the optimal solution? 

<b>Question 4 (b): </b>Could we assign a "hard clustering" of all points instead of a probability?


<b>Question 5:</b> Can a cluster in a hard EM clustering be empty? 

<b>Question 6:</b> Is the EM algorithm guaranteed to converge? 

HINT: Assume the objective is strictly decreasing between every step of the algorithm. 

<b>Question 7 (a): </b>How is Expectation Maximization Clustering a generalizaton of k-means clustering?

HINT: See the book [ZM] page 353. 

<b>Question 7 (b): </b>Both models rely on certain assumptions on the data. What are these assumptions? Which assumptions are shared?

<b>Question 7 (c): </b>Give examples of data where the assumptions of k-means clustering and EM clustering are reasonable/unreasonable. 

HINT: Draw data that satisfies/doesn't satisfy all assumptions for k-means clustering. What about EM clustering? 

<!--
TODO: Some question that I thought about which would be nice to have is like a discussion about the differences between Kmeans and EM. The book points out (page 353) that EM is some kind of generalization of Kmeans, so in some sense it's more versatil (for example, it doesn't restrict us to convex clusters -right?- and it seems like the fact that the clusters are ''randomized'' allows us to do more things). What are the advantages/disadvantages of each? -->


# EM-clustering as a Hidden Markov Model (optional)

One can also think of the EM-clustering model as a particularly simple HMM. If we consider a model with $k$ clusters, we can define a corresponding HMM with $k$ hidden states. The transition probability from any state $j$ to any state $i$ is identical to the prior probability, $\Pr(C_i)$. The emission probability in hidden state $i$ is $\Pr(x \ | \ C_i)$, (a gaussian probability distribution). 


The data set we have, $D =\{x_1, \dots x_m \} $ corresponds to the emissions from the HMM, and the (hard) clustering is the decoding! Since the transition probabilities $\Pr(z_n = i \ | \ z_{n-1} = j) = P(C_i)$ do not depend on $j$, finding a decoding is particularly simple. In fact, the Viterbi and the Posterior decoding will give the same results (why?). The result will be, that $z_j = \text{argmax}_i \Pr(x_j \ | \ C_i)$. 


Now, the Baum-Welch training we saw for HMMs summarized in slide 20 in the slide-set hmm-training is *exactly* the algorithm presented above for EM-clustering! You may try to derive this if you have the time.

# Expectation Maximization (code)
In this exercise you must implement the EM algorithm. To test our implementation we will need data. Like last week we use the Iris dataset. Recall that the dataset has three clases so we <i>cheat</i> by setting $k=3$.

In [1]:
# Load the Iris data set
import sklearn.datasets
iris = sklearn.datasets.load_iris()
X = iris['data'][:,0:2] # reduce dimensions so we can plot what happens, you may experiment with this.
k = 3
print(X.shape)

(150, 2)


## Implementing EM algorithm
Remember the Expectation Maximization algorithm has two steps. Let us first implement the Expectation step. For this step it suffices to compute $\Pr(C_i\mid x_j)$ for $i=1,\ldots,k$ and $j=1,\ldots n$. Remember that $\Pr(C_i \mid x_j)$ can be computed as 

$$\Pr(C_i|x_j) = \frac{\Pr(x_j|C_i)\cdot \Pr(C_i)}{\Pr(x_j)}$$

The first goal will then be to compute each part of the above equation. To do this we will need the parameters: ($\mu_i, \Sigma_i, \Pr(C_i)$). These will be represented in Python as follows: 

$$\text{means}=\begin{pmatrix}
- & \mu_1^T & - \\
- & \vdots & - \\
- & \mu_k^T & - 
\end{pmatrix}\in \mathbb{R}^{k\times d} \quad\quad
\text{probs_c}=\begin{pmatrix}
\Pr(C_1)\\
\vdots \\
\Pr(C_k)
\end{pmatrix}\in \mathbb{R}^{k}
$$

Similarly we represent the $\Sigma_i$'s as $\text{covs}\in\mathbb{R}^{k\times d \times d}$ such that $\text{covs[i]}=\Sigma_i$. Finally we represent $\Pr(x)$'s and $\Pr(C_i \mid x)$ as 

$$\text{prob_x}=\begin{pmatrix}
\Pr(x_1) \\
\vdots \\
\Pr(x_n) 
\end{pmatrix}\in \mathbb{R}^{n} \quad\quad
\text{probs_cx}=\begin{pmatrix}
\Pr(C_1\mid x_1) & \ldots & \Pr(C_1 \mid x_n)\\
\vdots \\
\Pr(C_k \mid x_1) & \ldots & \Pr(C_k \mid x_n)
\end{pmatrix}\in \mathbb{R}^{k\times n}
$$

The function *compute_probs_cs* below takes `means`, `probs_c` and `covs` as input and returns `probs_cx` and `probs_x`. 

<b>Question: </b>What is the dimensions of `probs_cx` and how can we compute it given `means`, `probs_c` and `covs`?

<!-- The algorithm returns a clustering $M = \{C_1,\ldots,C_k\}$. This corresponds to a 
 - sequence of means $(\mu_1,\ldots,\mu_k)$ where each $\mu_i\in\Bbb R^d$,
 - sequence of covariance matrices $(\Sigma_1,\ldots,\Sigma_k)$ where each $\Sigma_i\in\Bbb R^{d\times d}$,
 - prior probabilities $(\Pr(C_1),\ldots,\Pr(C_k))$. 

Given $x$ we can then compute the probability of $x$ conditioned on the $i$'th cluster $Pr(C\mid x_i)$. If we want to assign $x$ to a specific cluster we compute $(\Pr(x|C_1),\ldots,\Pr(x|C_k))$ and assign $x$ to the `arg max`.-->

<!--The following helper function takes a description of a Gaussian Mixture ($\mu_i$'s, $\Sigma_i$'s and $\Pr(C_i)$'s)) and returns the probability densities of each point. We represent the inputs as 

If you want more information you can take a look at scipy's [multivariate_normal](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.multivariate_normal.html) documentation.-->

In [36]:
import numpy as np
from scipy.stats import multivariate_normal

def compute_probs_cx(points, means, covs, probs_c):
    '''
    Input
      - points:    (n times d) array containing the dataset
      - means:     (k times d) array containing the k means
      - covs:      (k times d times d) array such that cov[j,:,:] is the covariance matrix of the j-th Gaussian.
      - probs_c:   (k) array containing priors
    Output
      - probs_cx:  (k times n) array such that the entry (i,j) represents Pr(C_i|x_j)
    '''
    # Convert to numpy arrays.
    points, means, covs, probs_c = np.asarray(points), np.asarray(means), np.asarray(covs), np.asarray(probs_c)
    
    # Get sizes
    n, d = points.shape
    k = means.shape[0]
    
    # Compute probabilities
    # This will be a (k, n) matrix where the (i,j)'th entry is Pr(C_i)*Pr(x_j|C_i).
    probs_cx = np.zeros((k, n))
    for i in range(k):
        
         # Handle numerical issues, these lines are unimportant for understanding the algorithm. 
        if np.allclose(np.linalg.det(covs[i]), 0):  # det(A)=0 <=> singular. 
            print("Numerical issues, run again. ") 
            return None, None
        
        probs_cx[i] = probs_c[i] * multivariate_normal.pdf(mean=means[i], cov=covs[i], x=points)
    
    # The sum of the j'th column of this matrix is P(x_j); why?
    probs_x = np.sum(probs_cx, axis=0, keepdims=True) 
    assert probs_x.shape == (1, n)
    
    # Divide the j'th column by P(x_j). The the (i,j)'th then 
    # becomes Pr(C_i)*Pr(x_j)|C_i)/Pr(x_j) = Pr(C_i|x_j)
    probs_cx = probs_cx / probs_x
    
    return probs_cx, probs_x

The following is the basic structure for the EM algorithm. For the Expectation step it calls the above code. You need to finish the Maximization step 

<b>Question: </b>Where in the code (and how) do we initialize `means`, `covs` and `probs_c`?

The above code computed the Expectation step. In the code below you should implement the Maximization step.

<b>Question: </b>Which three things do you need to compute in the Maximization step?

In [40]:
def em_algorithm(X, k, T, epsilon = 0.001, means=None):
    """ Clusters the data X into k clusters using the Expectation Maximization algorithm. 
    
        Parameters
        ----------
        X : Data matrix of shape (n, d)
        k : Number of clusters.
        T : Maximum number of iterations
        epsilon :  Stopping criteria for the EM algorithm. Stops if the means of
                   two consequtive iterations are less than epsilon.
        means : (k times d) array containing the k initial means (optional)
        
        Returns
        -------
        means:     (k, d) array containing the k means
        covs:      (k, d, d) array such that cov[j,:,:] is the covariance matrix of 
                   the Gaussian of the j-th cluster
        probs_c:   (k, ) containing the probability Pr[C_i] for i=0,...,k. 
        llh:       The log-likelihood of the clustering (this is the objective we want to maximize)
    """
    n, d = X.shape
    
    # Initialize and validate mean
    if means is None: 
        means = np.random.rand(k, d)

    # Initialize cov, prior
    probs_x  = np.zeros(n) 
    probs_cx = np.zeros((k, n)) 
    probs_c  = np.zeros(k) + np.random.rand(k)
    
    covs = np.zeros((k, d, d))
    for i in range(k): covs[i] = np.identity(d)
    probs_c = np.ones(k) / k
    
    # Column names
    print("Iterations\tLLH")
    
    close = False
    old_means = np.zeros_like(means)
    iterations = 0
    while not(close) and iterations < T:
        old_means[:] = means 

        # Expectation step
        probs_cx, probs_x = compute_probs_cx(X, means, covs, probs_c)
        if probs_cx is None: return em_algorithm(X, k, T, epsilon = epsilon)
        assert probs_cx.shape == (k, n)
        
        # Maximization step
        # YOUR CODE HERE
        for i in range(k): # loop over clusters
            probs_c[i] = 1 / n * np.sum(probs_cx[i, :])
            means[i] = probs_cx[i, :] @ X / np.sum(probs_cx[i, :])
            for j in range(n):
                covs[i] += np.outer(X[j] - means[i], X[j] - means[i]) * probs_cx[i, j]
            covs[i] /= np.sum(probs_cx[i, :])
        # END CODE
        
        # Compute per-sample average log likelihood (llh) of this iteration     
        llh = 1/n*np.sum(np.log(probs_x))
        print(iterations+1, "\t\t", llh)

        # Stop condition
        dist = np.sqrt(((means - old_means) ** 2).sum(axis=1))
        close = np.all(dist < epsilon)
        iterations += 1
        
    # Validate output
    assert means.shape == (k, d)
    assert covs.shape == (k, d, d)
    assert probs_c.shape == (k,)
    return means, covs, probs_c, llh

means, covs, priors, llh = em_algorithm(X, 3, 100, 0.001)

Iterations	LLH
1 		 -18.381793170547887
2 		 -1.8030235593626902
3 		 -1.7994581953907764
4 		 -1.7980275055931885
5 		 -1.7959514209006322
6 		 -1.7924234136353134
7 		 -1.7865397730627237
8 		 -1.7772320256966123
9 		 -1.7633128524059465
10 		 -1.7449817646400767
11 		 -1.727265538404193
12 		 -1.715367766394926
13 		 -1.707946845601865
14 		 -1.704079568318205
15 		 -1.7021989891303277
16 		 -1.700855000752647
17 		 -1.699554048831472
18 		 -1.6979496364034135
[[inf inf]
 [inf inf]]
nan
19 		 -1.695668308456152
[[nan nan]
 [nan nan]]
nan
20 		 -1.6921441293329222
[[nan nan]
 [nan nan]]
nan
21 		 -1.6865975773772188
[[nan nan]
 [nan nan]]
nan
22 		 -1.6789501464445047
[[nan nan]
 [nan nan]]
nan
23 		 -1.6705205684630309
[[nan nan]
 [nan nan]]
nan
24 		 -1.6619841738502636
[[nan nan]
 [nan nan]]
nan
25 		 -1.6530393657448725
[[nan nan]
 [nan nan]]
nan
26 		 -1.6430876422711898
[[nan nan]
 [nan nan]]
nan
27 		 -1.6313184874626165
[[nan nan]
 [nan nan]]
nan
28 		 -1.6166526678692696
[[n

/Users/au597881/anaconda3/lib/python3.7/site-packages/numpy/linalg/linalg.py:2022: RuntimeWarning: overflow encountered in det
  r = _umath_linalg.det(a, signature=signature)
/Users/au597881/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:57: RuntimeWarning: overflow encountered in true_divide
/Users/au597881/anaconda3/lib/python3.7/site-packages/numpy/linalg/linalg.py:2022: RuntimeWarning: invalid value encountered in det
  r = _umath_linalg.det(a, signature=signature)
/Users/au597881/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:54: RuntimeWarning: invalid value encountered in true_divide
/Users/au597881/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:66: RuntimeWarning: invalid value encountered in less


31 		 -1.541132129158364
[[nan nan]
 [nan nan]]
nan
32 		 -1.5167781443765107
[[nan nan]
 [nan nan]]
nan
33 		 -1.5081309357147934
[[nan nan]
 [nan nan]]
nan
34 		 -1.5069580332464585
[[nan nan]
 [nan nan]]
nan
35 		 -1.5068613269462308
[[nan nan]
 [nan nan]]
nan
36 		 -1.506853605138667
[[nan nan]
 [nan nan]]
nan
37 		 -1.5068532667043077
[[nan nan]
 [nan nan]]
nan
38 		 -1.5068533818724832
[[nan nan]
 [nan nan]]
nan
39 		 -1.5068534413720853
[[nan nan]
 [nan nan]]
nan
40 		 -1.506853461984685
[[nan nan]
 [nan nan]]
nan
41 		 -1.5068534684123642
[[nan nan]
 [nan nan]]
nan
42 		 -1.506853470339518
[[nan nan]
 [nan nan]]
nan
43 		 -1.5068534709078323
[[nan nan]
 [nan nan]]
nan
44 		 -1.5068534710741976
[[nan nan]
 [nan nan]]
nan
45 		 -1.5068534711227348
[[nan nan]
 [nan nan]]
nan
46 		 -1.5068534711368733
[[nan nan]
 [nan nan]]
nan
47 		 -1.506853471140989
[[nan nan]
 [nan nan]]
nan
48 		 -1.5068534711421866
[[nan nan]
 [nan nan]]
nan
49 		 -1.5068534711425348
[[nan nan]
 [nan nan]]
na

The random initialization usually causes the algorithm to get stuck at different local maxima. This causes different runs to get different scores. In practice this can for example be handled by running the algorithm several times and picking the best run. 

The following code runs EM algorithm 50 times and plots the score of each run. Because the data set is fairly small, $n=150$, most of the runs will get almost the same score. 

In [ ]:
%matplotlib notebook 
import matplotlib.pyplot as plt

fig, ax = plt.subplots(1, 1, figsize=(8,3))    
llhs = []

for i in range(100):
    _, _, _, llh = em_algorithm(X, 3, 100)
    llhs.append(llh)
    ax.plot(llhs, 'bx')
    fig.canvas.draw() 

To check your implementation you should run <a href="http://scikit-learn.org/stable/modules/mixture.html">sklearn</a>'s implementation of the EM algorithm. You might want to take a look at the documentation to get a better understanding of what the algorithm is actually doing.

<!--By default the implementation repeats the algorithm $10$ times and picks the best result. A sanity check for your implementation of Lloyd's algorithm is to check that the scores are roughly the same. -->

In [ ]:
from sklearn.mixture import GaussianMixture as EM
expectation_maximization = EM(n_components=3, init_params='random', covariance_type='diag', verbose=2, verbose_interval =1).fit(X)

print(expectation_maximization.score(X))

To get a visual understanding of the algorithm, the following code visualizes each step of the algorithm. Similarly to previous week, it prints the centroids and the corresponding cluster each point is assigned to. Furthermore, because the clusters are represented as gaussians, the gaussians are plotted. They are first plotted by themselves, and then they are plotted together (by summing). 

To run the visualization you need to copy and paste your implementation from above. Then the code should run with no modifications. Running the visualization might be a bit slow; try change the `detail` parameter if you have any problems. 

In [ ]:
%matplotlib notebook 
import matplotlib.pyplot as plt
import numpy as np
import time
from mpl_toolkits.mplot3d import Axes3D

def em_algorithm_visualize(X, k, T, epsilon = 0.001, means=None, detail=20):
    """ Clusters the data X into k clusters using the Expectation Maximization algorithm. 
    
        Parameters
        ----------
        X : Data matrix of shape (n, d)
        k : Number of clusters.
        T : Maximum number of iterations
        epsilon :  Stopping criteria for the EM algorithm. Stops if the means of
                   two consequtive iterations are less than epsilon.
        means : (k times d) array containing the k initial means (optional)
        
        Returns
        -------
        means:     (k, d) array containing the k means
        covs:      (k, d, d) array such that cov[j,:,:] is the covariance matrix of 
                   the Gaussian of the j-th cluster
        probs_c:   (k, ) containing the probability Pr[C_i] for i=0,...,k. 
        llh:       The log-likelihood of the clustering (this is the objective we want to maximize)
    """
    n, d = X.shape
    
    # Visualization stuff. 
    fig, (ax, _, _, _, _) = plt.subplots(5, 1, figsize=(10,16)) 
    ax.axis('off')
    colors = ["r", "g", "b"]
    ax3d = fig.add_subplot(2, 1, 2, projection='3d')
    ax3d1 = fig.add_subplot(3, 1, 2, projection='3d')
    ax3d2 = fig.add_subplot(4, 1, 2, projection='3d')
    ax3d3 = fig.add_subplot(5, 1, 2, projection='3d')
    ax3d.set_axis_off()
    ax3d1.set_axis_off()
    ax3d2.set_axis_off()
    ax3d3.set_axis_off()
    
    # Initialize and validate mean
    if means is None: means = np.random.rand(k, d)

    # Initialize 
    probs_x  = np.zeros(n) 
    probs_cx = np.zeros((k, n)) 
    probs_c  = np.zeros(k) 
    covs = np.zeros((k, d, d))
    for i in range(k): covs[i] = np.identity(d)
    probs_c = np.ones(k) / k

    # END CODE
    
    # Column names
    print("Iterations\tLLH")
    
    close = False
    old_means = np.zeros_like(means)
    iterations = 0
    while not(close) and iterations < T:
        old_means[:] = means # This allows us to actually copy the array mean

        # Expectation step
        probs_cx, probs_x = compute_probs_cx(X, means, covs, probs_c)
        if probs_cx is None: return em_algorithm(X, k, T, epsilon = epsilon)
        assert probs_cx.shape == (k, n)
        
        # Maximization step
        # YOUR CODE HERE
        for i in range(k): # loop over clusters
            probs_c[i] = 1 / n * np.sum(probs_cx[i, :])
            means[i] = probs_cx[i, :] @ X / np.sum(probs_cx[i, :])
            for j in range(n):
                covs[i] += np.outer(X[j] - means[i], X[j] - means[i]) * probs_cx[i, j]
            covs[i] /= np.sum(probs_cx[i])
        # END CODE
        
        # Finish condition
        dist = np.sqrt(((means - old_means) ** 2).sum(axis=1))
        close = np.all(dist < epsilon)
        iterations += 1
        
        
        # !----------- VISUALIZATION CODE -----------!
        centroids = means
        # probs_cx's (i,j) is Pr[C_i, x_j]
        # assign each x_i to the cluster C_i that maximizes P(C_i | x_j)
        clustering = np.argmax(probs_cx, axis=0)
        assert clustering.shape == (n,), clustering.shape
        
        # Draw clusters
        ax.cla()
        for j in range(k):
            centroid = centroids[j]
            c = colors[j]
            ax.scatter(centroid[0], centroid[1], s=123, c=c, marker='^')
            data = X[clustering==j]
            x = data[:,0]
            y = data[:,1]
            ax.scatter(x, y, s=3, c=c)
            
        # draw 3d gaussians. 
        #Create grid and multivariate normal
        xs = np.linspace(4,7, 50)
        ys = np.linspace(2,4.5, 50)
        Xs, Ys = np.meshgrid(xs, ys)
        pos = np.empty(Xs.shape + (2,))
        pos[:, :, 0] = Xs; pos[:, :, 1] = Ys
        prob_space = sum([multivariate_normal(means[j], covs[j]).pdf(pos) for j in range(k)])

        #Make a 3D plot
        ax3d.cla()
        ax3d1.cla()
        ax3d2.cla()
        ax3d3.cla()
        ax3d.plot_surface(Xs, Ys, prob_space, cmap='viridis', linewidth=0)
        ax3d1.plot_surface(Xs, Ys, multivariate_normal(means[0], covs[0]).pdf(pos), cmap='viridis', linewidth=0)
        ax3d2.plot_surface(Xs, Ys, multivariate_normal(means[1], covs[1]).pdf(pos), cmap='viridis', linewidth=0)
        ax3d3.plot_surface(Xs, Ys, multivariate_normal(means[2], covs[2]).pdf(pos), cmap='viridis', linewidth=0)
        
        fig.canvas.draw()
        
    # Validate output
    assert means.shape == (k, d)
    assert covs.shape == (k, d, d)
    assert probs_c.shape == (k,)
    return means, covs, probs_c, llhs

# Load the Iris data set
import sklearn.datasets
iris = sklearn.datasets.load_iris()
X = iris['data'][:,0:2] # reduce dimensions so we can plot what happens.
k = 3

# the higher the detail the slower plotting
detail = 20 # 50 looks very nice but your computer might not be able to handle it. 
means, covs, priors, llh = em_algorithm_visualize(X, 3, 40, 0.001, detail=detail)

## Initializing EM with Lloyd's algorithm 

So far we have been initializing the means for the Expectation Maximization algorithm at random. We could also make a mix of Lloyd's algorithm and EM algorithm by running Lloyd's algorithm first to obtain the initial means for the EM algorithm. 

Begin by copying and pasting the implementation of Lloyd's algorithm from the previous week

In [ ]:
def lloyds_algorithm(X, k, T):
    """ Clusters the data of X into k clusters using T iterations of Lloyd's algorithm. 
    
        Parameters
        ----------
        X : Data matrix of shape (n, d)
        k : Number of clusters.
        T : Maximum number of iterations to run Lloyd's algorithm. 
        
        Returns
        -------
        clustering: A vector of shape (n, ) where the i'th entry holds the cluster of X[i].
        centroids:  The centroids/average points of each cluster. 
        cost:       The cost of the clustering 
    """
    pass

Then use this algorithm to initialize the means for the EM algorithm. For this notice that `em_algorithm` accepts an optional input for the initial means.

Also, notice that the sklearn's implementation of the EM algorithm can take this initialization into account. Can you look through the documentation and find out what lines should be changed when we used sklearn before? This would be very useful for testing and comparing your implementation above.

In [ ]:
# YOUR CODE HERE
# END CODE

<b>Question:</b> Why can we use Lloyd's algorithm to initialize the EM algorithm? Does it (always) give a better final clustering? If so, why is that the case?

# Clustering digits
Last week, we performed clustering of digits by the use of Lloyds algorithm. Now, we try the same with the EM clustering algorithm. The following code imports the data set:


In [ ]:
import torchvision.datasets as datasets
mnist_trainset = datasets.MNIST(root='./data', train=True, download=True, transform=None)

X = mnist_trainset.train_data.numpy().reshape((60000, 28*28))
y = mnist_trainset.train_labels.numpy()

print(X.shape)

The following code runs the Expectation Maimization algorithm on 5000 images from the MNIST dataset. It then visualizes the found centroids. 

In [ ]:
from sklearn.mixture import GaussianMixture as EM
import matplotlib.pyplot as plt

# One cluster for each digit
k = 10

# Run EM algorithm on 1000 images from the MNIST dataset. 

expectation_maximization = EM(n_components=k, max_iter=10, init_params='kmeans', covariance_type='diag', verbose=1, verbose_interval =1).fit(X)
print(expectation_maximization.score(X[:1000]))

means = expectation_maximization.means_
covs = expectation_maximization.covariances_
      
print(means.shape)
fig, ax = plt.subplots(1, k, figsize=(8, 1))

for i in range(k):
    ax[i].imshow(means[i].reshape(28, 28), cmap='gray')
    
plt.show()

<b>Question 1: </b>Why do the centroids look like images of digits? 
    
A particularly cool thing about representing a clustering as probability distributions is that it is possible to sample points from them! This way you can generate *new* images of digits. A similar idea lies behind some of the new cool techniques for generation of images such as variational autoencoders. Lets try sampling!

We have implemented a function to do this for you:

In [ ]:
from scipy.stats import multivariate_normal
import numpy as np

def sample(means, covs, num):
    mean = means[num]
    cov = covs[num]
     
    fig, ax = plt.subplots(1, 10, figsize=(8, 1))
    
    for i in range(10):
        img = multivariate_normal.rvs(mean=mean, cov=cov) # draw random sample   
        ax[i].imshow(img.reshape(28, 28), cmap='gray') # draw the random sample
    plt.show()

sample(means, covs, 5)  ##Try changing the number